### DISCUSSION TITLE
Refresh rate sync to fps

### DISCUSSION THREAD

**Discussion by jorgebef** [wezterm/wezterm#2419] [Q&A]
Hi all,

I've been using Kitty terminal for quite some time now and I discovered wezterm by chance (reddit) and I quite like it, however, I can't switch to it completely due to some heavy "lag" or out of sync from the terminal refresh rate and my 4k monitor (which I use as main display).

I currently use a m1 pro Macbook Pro with the pro motion display, and when I use wezterm on that display, it is smooth as butter and absolutely stunning.
The problem is when I use wezterm on the external monitor (4k @ 60Hz), which stutters and has a noticeable delay when pressing keys.

I feel like the terminal is refreshing at a higher speed than the screen can display, therefore it looks like it's lagging, but it's just how it looks since the commands all go through correctly (if that makes sense).

Maybe it would be convenient to include some sort of option to limit the refresh rate or similar performance tweaking options like Kitty has?
Link for reference: [Kitty Performance options Docs](https://sw.kovidgoyal.net/kitty/performance/)

It also could be some issue on my part, if anyone with a similar setup is using it with no issue I would love to hear some feedback.

Here are two videos showing the difference (it's more noticeable as the lag is felt between the key press and the movement on screen).

Kitty- 4k @ 60Hz
https://imgur.com/a/k3bPj3s

Wezterm - 4K @ 60Hz
https://imgur.com/a/4fIyONJ

**Comment by wez** [comment]
8ca3faa6e hooks up the existing undocumented `max_fps` config option on macOS. `max_fps` defaults to 60.

One thing to note about your test scenario is that scrolling a vertically split screen in vim is a relatively expensive `O(screen-area)` operation.  It's possible that the bottleneck isn't the frame rate so much as it is the cost of updating that area, so I'm curious to hear how much of an impact this change has on your test scenario.

That change should be available in a nightly download in about an hour from now.

**Reply by jorgebef** [reply]
Hi, thanks for the hint, it made a difference and the option was indeed working and the difference it can make is noticeable.

However, I have now gotten to the next step (lol).

When the wezterm window is as large as the screen (maximized) the scrolling and in general any other operation that requires quick redrawing (not only in neovim) becomes quite laggy.
However, if I open 3 wezterm windows, each of them being 1/3 or the screen split along from left to right, then they are perfectly fine, the scroll is smooth as butter and all other operations too.

I guess this is due to the fact that wezterm redraws the whole output and I have no clue whether there is any possible fix.

Also, if any logs need to be posted or if I can be of assistance, I'm more than happy to help.

My machine is an M1 pro Macbook Pro 16", in case any other tests for other issues could be tested on my machine and you don't have one available with the exact same specs. I would love to be able to contribute in any way to the project.

Thanks again and looking forward to your input.

**Reply by wez** [reply]
I landed some changes today that may help with this. Give it an hour or so for `main` to build out before trying the latest nightly build and let me know how it goes.

**Reply by jorgebef** [reply]
I can say there has been a noticeable improvement when (as mentioned in my previous comment) I reduce the size of the terminal to a vertical third of the screen. Doing that, it is butter smooth setting max_fps=240 (which seems the maximum acceptable value).

Unfortunately, in fullscreen, it doesn't reach a level where it is smooth enough.
